### PR TITLE
nspawn: refuse to bind mount device node from host when --private-users= is specified

### LINKED ISSUE #34243: systemd-nspawn: /dev/net/tun permission take over
### systemd version the issue has been seen with

256.5-1

### Used distribution

Arch Linux

### Linux kernel version used

6.10.7

### CPU architectures issue was seen on

x86_64

### Component

systemd-nspawn

### Expected behaviour you didn't see

0 crw-rw-rw- 1 root root 10, 200 Sep  3 21:10 /dev/net/tun

### Unexpected behaviour you saw

0 crw-rw-rw- 1 vu-container-0 vg-container-0 10, 200 Sep  3 21:10 /dev/net/tun

Everytime you start a container the /`dev/net/tun` device got the permission of the unprivileged user (root) from container. But you have to manual `chown root: /dev/net/tun` to make /dev/net/tun work for example with qemu. This does **not** happen for example with other device like `/dev/dri/card0`

### Steps to reproduce the problem

Create nspawn file: /etc/systemd/nspawn/container.nspawn

```
[Exec]
Ephemeral=on
Boot=yes

NoNewPrivileges=yes
PrivateUsers=pick
Hostname=container

[Network]
Private=yes
VirtualEthernet=yes
```

Start via `systemctl start systemd-nspawn@container`

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #34243.

### CODE REVIEW

File: src/nspawn/nspawn.c [hunk 1]
```diff
@@ -2194,7 +2194,7 @@ static int copy_devnodes(const char *dest) {
                                 /* Explicitly warn the user when /dev is already populated. */
                                 if (errno == EEXIST)
                                         log_notice("%s/dev/ is pre-mounted and pre-populated. If a pre-mounted /dev/ is provided it needs to be an unpopulated file system.", dest);
-                                if (errno != EPERM)
+                                if (errno != EPERM || arg_uid_shift != 0)
```
Comment (poettering) [inline src/nspawn/nspawn.c]: maybe switch to ERRNO_IS_PRIVILEGE() for the EPERM check if you touch this?

**Comment by yuwata** [comment]
@poettering Thank you for the review. I also made the following `lchown()` is called only when `mknod()` succeeds. Please review again.

**Comment by helmutg** [comment]
This PR made it into Debian bookworm via a stable update. A `systemd-nspawn` run from a custom `.service` unit with `DevicePolicy=closed` and no `DeviceAllow` would formerly work in the sense that `/dev/net/tun` could neither be bind mounted (as it was hidden by the device policy) nor created (as `mknod` failed) and thus would run the container without this device. Due to this change, the `mknod` failure became fatal and containers now fail to start. Adding `DeviceAllow` for `/dev/net/tun` fixes containers (for details see the discussion of the underlying bug).

It is unclear whether this is to be considered a regression (containers that formerly worked now fail) or whether this is a misconfiguration (failure to add `/dev/net/tun` to `DeviceAllow`).

In principle, I would that if passing `--private-users` and no `--private-network`, then `/dev/net/tun` can never be used in the container and thus should not be created. Do others consider this a sufficiently severe bug to warrant a regression update?

**Comment by yuwata** [comment]
@helmutg Thank you for the report. Filed the potential regression as #35116, to make your issue not forgotten. I will consider it later.

**Comment by yuwata** [comment]
@helmutg Fix is waiting in #35116.

**Pull request merged**